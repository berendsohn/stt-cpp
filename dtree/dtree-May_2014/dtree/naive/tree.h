// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_TREE_H_
#define DTREE_NAIVE_TREE_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndTree : public Base {
 public:
  EndTree() : parent_(NULL) {}
  EndTree* parent() const { return parent_; }
  void set_parent(EndTree* parent) { parent_ = parent; }
  template<typename Forwarder>
  void CopyFrom(const EndTree& node, Forwarder forward);
 private:
  EndTree* parent_;
  DISALLOW_COPY_AND_ASSIGN(EndTree);
};

template<typename Type, typename Base>
class WithStaticValue : public Base {
 public:
  typedef dtree::Nop<Type> Group_;
  Type value() const { return value_; }
  void set_value(Type value) { value_ = value; }
  //
  template<typename Node>
  static Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, Type value);
 protected:
  WithStaticValue() : value_() {}
 private:
  Type value_;
};

template<typename Group, typename Base>
class WithAncValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  //
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void AddToProperAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromProperAnc(Node* node, typename Group::Type delta);

 protected:
  WithAncValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAncAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  //
  template<typename Node>
  static typename Semigroup::Type AggrAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostAnc(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostAnc(Node* node, const Predicate& predicate);
  template<typename Node>
  static typename Semigroup::Type AggrProperAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostProperAnc(int dir,
                                    Node* node,
                                    const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostProperAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostProperAnc(Node* node, const Predicate& predicate);

 protected:
  WithAncAggr() {}
};

template<typename Base>
class WithEvertBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvertBy() {}
};

template<typename Base>
class WithEvert : public Base {
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvert() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_TREE_H_
