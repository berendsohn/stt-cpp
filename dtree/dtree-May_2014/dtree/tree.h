// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_TREE_H_
#define DTREE_TREE_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/tree.h"
#else
#include "dtree/selfadjust/tree.h"
#endif
#endif  // DTREE_TREE_H_
