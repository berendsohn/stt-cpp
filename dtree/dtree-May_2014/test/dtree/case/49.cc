#define DTREE_CONFIG EndTreeWithDesc<WithAncAggr<FreeSemigroup, NullPredicate_, WithAncAggr<FreeSemigroup, NullPredicate_, WithDescValueWithDesc<FreeGroup, WithAncDescAggr<FreeSemigroup, NullPredicate_, WithAncDescAggr<FreeSemigroup, NullPredicate_, WithAncDescValueWithDesc<FreeGroup, Begin > > > > > > >
#include "test/dtree/tree.cc"
