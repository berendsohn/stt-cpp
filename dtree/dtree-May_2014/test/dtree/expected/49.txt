<<<
  1: v507,000,a0d,000 vc10,000,988,000
  2: v84c,000,6f8,000 v35e,000,498,000
  3: va2a,000,89e,000 v57f,000,c75,000 -> 2
  4: v8e8,000,199,000 ve37,000,d0b,000 -> 5
  5: v54b,000,081,000 ve09,000,cd5,000 -> 13
  6: vb50,000,e86,000 v674,000,c38,000 -> 4
  7: v59c,000,757,000 v6d0,000,89d,000
  8: v942,000,290,000 ve98,000,d45,000 -> 4
  9: va03,000,d17,000 va43,000,db7,000
  10: v00d,000,600,000 v399,000,29e,000 -> 7
  11: vcf9,000,571,000 vecd,000,810,000
  12: va0a,000,fcf,000 v75c,000,812,000 -> 7
  13: v13b,000,cd2,000 v8fd,000,12c,000 -> 2
  14: v135,000,a57,000 v0bf,000,a9e,000 -> 13
  15: v248,000,462,000 vfdc,000,123,000
>>>
<<<
  @1: v507,000,a0d,000 v507,000,a0d,000 a000,000 o000,000 v507,000,a0d,000 v507,000,a0d,000 a000,000 o000,000 a000,000 o000,000 vc10,000,988,000 a000,000 a000,000

    /2: v84c,a2a,6f8,89e v84c,a2a,6f8,89e a84c,6f8 o000,6f8 v84c,a2a,6f8,89e v84c,a2a,6f8,89e a84c,6f8 o000,6f8 a84c,6f8 o000,6f8 v35e,57f,498,c75 a35e,498 a35e,498
     ->13: v13b,000,cd2,6f8 v13b,000,cd2,6f8 a000,cd2 o000,cd2 v13b,000,cd2,6f8 v13b,000,cd2,6f8 a000,cd2 o000,cd2 a000,cd2 o000,cd2 v8fd,35e,12c,498 a8fd,12c a8fd,12c
       ->14: v135,000,a57,cd2 v135,000,a57,cd2 a000,a57 o000,a57 v135,000,a57,cd2 v135,000,a57,cd2 a000,a57 o000,a57 a000,a57 o000,a57 v0bf,8fd,a9e,12c a0bf,a9e a0bf,a9e
        \5: v54b,13b,081,cd2 v54b,13b,081,cd2 a54b,081 o000,081 v54b,13b,081,cd2 v54b,13b,081,cd2 a54b,081 o000,081 a54b,081 o000,081 ve09,8fd,cd5,12c ae09,cd5 ae09,cd5
          \4: v8e8,54b,199,081 v8e8,54b,199,081 a8e8,199 o000,199 v8e8,54b,199,081 v8e8,54b,199,081 a8e8,199 o000,199 a8e8,199 o000,199 ve37,e09,d0b,cd5 ae37,d0b ae37,d0b
           ->6: vb50,000,e86,199 vb50,000,e86,199 a000,e86 o000,e86 vb50,000,e86,199 vb50,000,e86,199 a000,e86 o000,e86 a000,e86 o000,e86 v674,e37,c38,d0b a674,c38 a674,c38
            \8: v942,8e8,290,199 v942,8e8,290,199 a942,290 o000,000 v942,8e8,290,199 v942,8e8,290,199 a942,290 o000,000 a942,290 o000,000 ve98,e37,d45,d0b ae98,d45 ae98,d45
  @3: va2a,000,89e,000 va2a,000,89e,000 a000,89e o000,89e va2a,000,89e,000 va2a,000,89e,000 a000,89e o000,89e a000,89e o000,89e v57f,000,c75,000 a57f,c75 a57f,c75

  @7: v59c,000,757,000 v59c,000,757,000 a000,000 o000,000 v59c,000,757,000 v59c,000,757,000 a000,000 o000,000 a000,000 o000,000 v6d0,000,89d,000 a000,000 a000,000
   ->12: va0a,000,fcf,757 va0a,000,fcf,757 a000,000 o000,000 va0a,000,fcf,757 va0a,000,fcf,757 a000,000 o000,000 a000,000 o000,000 v75c,6d0,812,89d a000,000 a000,000
    \10: v00d,59c,600,757 v00d,59c,600,757 a000,000 o000,000 v00d,59c,600,757 v00d,59c,600,757 a000,000 o000,000 a000,000 o000,000 v399,6d0,29e,89d a000,000 a000,000

  @9: va03,000,d17,000 va03,000,d17,000 a000,000 o000,000 va03,000,d17,000 va03,000,d17,000 a000,000 o000,000 a000,000 o000,000 va43,000,db7,000 a000,000 a000,000

  @11: vcf9,000,571,000 vcf9,000,571,000 a000,000 o000,000 vcf9,000,571,000 vcf9,000,571,000 a000,000 o000,000 a000,000 o000,000 vecd,000,810,000 a000,000 a000,000

  @15: v248,000,462,000 v248,000,462,000 a000,000 o000,000 v248,000,462,000 v248,000,462,000 a000,000 o000,000 a000,000 o000,000 vfdc,000,123,000 a000,000 a000,000
>>>
